### ISSUE TITLE
Square bracket loops and foreach loops

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#22638]
We defined the behavior of square bracket loops like `[i in expr] body` before we had introduced the `foreach` loop to Chapel.  As a result, we defined it as being equivalent to `forall i in expr do body` if `expr` supported a parallel iterator, and like the serial loop `for i in expr do body` otherwise.

Now that we have `foreach` loops, it seems as though `foreach` would be the more appropraite fallback choice to use in the documentation and implementation.  Specifically, using the square-bracket loop form's interpretation as a "parallel if possible" loop suggests it's reasonable to execute in parallel, so serializing it feels like overkill if using a foreach would result in vectorization of GPU-ization of the loop.

**Comment by bradcray** [comment]
I'm particularly curious for reactions to this from @vasslitvinov (who was involved in the formulation and revised implementation of square bracket loops) and @mppf and @e-kayrakli (since they were involved in introducing `foreach` as well as vectorization and GPUization).

**Comment by mppf** [comment]
I agree that it would make sense for `[i in expr] body` to mean `foreach` or `forall`, depending on what iterators are available. If the author of this loop wants it to be fully serial, they should be writing `for i in expr do body`.

**Comment by vasslitvinov** [comment]
I agree, too.

If the []-loop is on a GPU locale, should parallel iterators still be considered?

**Comment by bradcray** [comment]
Since we support forall loops on GPU locales, I'd expect lowering it to a forall loop, if appropriate, to be a reasonable thing to do.

I'm going to put a 2.0 label on here because I think we should at least update the documentation before 2.0—though if it were simple to update the implementation as well, I think that would be great.

**Comment by e-kayrakli** [comment]
I agree with using `foreach` as the fallback, and with parallel iterator behavior being the same on a GPU locale.

**Comment by bradcray** [comment]
@vasslitvinov : I'm looking for the release (or better, PR) that introduced the interpretation of square brackets as "forall if possible, otherwise for" and coming up short.  Do you recall when/where that was?

**Comment by vasslitvinov** [comment]
> 'm looking for the release (or better, PR) that introduced the interpretation of square brackets as "forall if possible, otherwise for"

We announced it in 1.19. One of the key PRs that implemented the switch is #12211. The design discussion took place in #11819.

**Comment by lydia-duncan** [comment]
It seems like there's agreement here - does this need an owner to implement it?

**Comment by bradcray** [comment]
Yes, I think so, where by "owner" I think we're talking about "someone to update the docs" in the short-term (but bonus points if they also coax the compiler into generating the foreach.  I looked into doing the implementation bit briefly after filing it, but my compiler chops aren't good enough these days to get it done in short order).

**Comment by daviditen** [comment]
I did some digging into what happens with square bracket loops. When parallel iterators are not available for the iterable expression, function resolution resolves them into a forall loop over the serial iterator. Later, lowerIterators lowers the forall loop to a C for loop and since it came from a forall, the orderIndependent flag is set to true. This effectively makes it a foreach loop as things already stand.

**Comment by bradcray** [comment]
That's good to hear, though it still feels a little unfortunate in a few respects:
* When we redefined square brackets as "use parallel iterators if possible, fall back to serial if not", we also defined that 'forall' should mean "use parallel iterators or generate an error if you cannot".  So it feels potentially confusing, or like a mode mismatch, to see a forall loop invoking a serial iterator in the IR where a foreach would be more natural.
* Similarly, if we were to do foreach-specific optimizations or transformations going forward, we run the risk of missing these because the pattern doesn't match.

What would it take to have functino resolution convert into a foreach rather than a forall?  It sounds like you've found the right place to do so?

**Comment by daviditen** [comment]
The place where it resolves the iterator and chooses a parallel one if it exists and a serial one if not is in `resolveForallHeader()` in compiler/AST/foralls.cpp. I think that would be the place to change it by doing something like removing the old forall and replacing it with a new foreach.
